File: cpp/non-linear-structures/hash.cpp
// Luis Espino (c) 2024

#include <iostream>

using namespace std;

class Hash
{
    public:
    int n, m;
    int *h;
    int min, max;
    
    Hash(int m, int min, int max)
    {
        this->m = m;
        this->min = min;
        this->max = max;
        init();
	}
    
    int division(int k)
    { 
        return (k % m); 
    }
    
    int linear(int k)
    { 
        return ((k+1) % m); 
    }
    
    void init()
    {
        n = 0;
        h = new int[m];
        for(int i=0; i<m; i++) 
            h[i] = -1;		
	}
    
    void insert(int k)
    {
        int i = division(k);
        while (h[i] != -1)
            i = linear (i);
        h[i] = k;
        n++;
        rehashing();
    }
    
    void rehashing()
    {
        if ((n*100/m)>=max)
        {
            //array copy
            int *temp = h;
            print();
            //rehashing
            int mprev = m;
            m = n*100/min;
            init();
            for (int i=0; i<mprev; i++)
                if (temp[i]!=-1)
                    insert(temp[i]);
        }
        else print();
    }
    
    void print()
    {
        cout << "[" ;
        for(int i=0; i<m; i++)
            cout << " " << h[i];
        cout << " ] " << (n*100/m) << "%" << endl;
    }
};

int main() {
    Hash *hash = new Hash(5, 20, 80); // m, min, max
    hash->insert(5);
    hash->insert(10);
    hash->insert(15);
    hash->insert(20);
    hash->insert(25);
    hash->insert(30);
    return 0;
}

File: 13-hast-table/pascal/hash.pas
program HashExample;

{$mode objfpc}{$H+}

uses
    SysUtils;

type
    { THash }
    THash = class
    private
        FCount, FSize: Integer;
        FMin, FMax: Integer;
        FData: array of Integer;
        function Division(k: Integer): Integer;
        function Linear(k: Integer): Integer;
        procedure Init;
        procedure Rehashing;
    public
        constructor Create(m, min, max: Integer);
        procedure Insert(k: Integer);
        procedure Print;
    end;

{ THash }

constructor THash.Create(m, min, max: Integer);
begin
    FSize := m;
    FMin := min;
    FMax := max;
    Init;
end;

function THash.Division(k: Integer): Integer;
begin
    Result := k mod FSize;
end;

function THash.Linear(k: Integer): Integer;
begin
    Result := (k + 1) mod FSize;
end;

procedure THash.Init;
var
    i: Integer;
begin
    FCount := 0;
    SetLength(FData, FSize);
    for i := 0 to FSize - 1 do
        FData[i] := -1;
end;

procedure THash.Insert(k: Integer);
var
    i: Integer;
begin
    i := Division(k);
    while FData[i] <> -1 do
        i := Linear(i);
    FData[i] := k;
    Inc(FCount);
    Rehashing;
end;

procedure THash.Rehashing;
var
    temp: array of Integer;
    i, prevSize: Integer;
begin
    if (FCount * 100 div FSize) >= FMax then
    begin
        SetLength(temp, FSize);
        for i := 0 to FSize - 1 do
            temp[i] := FData[i];

        Print;

        prevSize := FSize;
        FSize := FCount * 100 div FMin;
        Init;

        for i := 0 to prevSize - 1 do
            if temp[i] <> -1 then
                Insert(temp[i]);
    end
    else
        Print;
end;

procedure THash.Print;
var
    i: Integer;
begin
    Write('[');
    for i := 0 to FSize - 1 do
        Write(' ', FData[i]);
    WriteLn(' ] ', (FCount * 100 div FSize), '%');
end;

{ === Programa principal === }

var
    hash: THash;

begin
    hash := THash.Create(5, 20, 80); // m, min, max
    hash.Insert(5);
    hash.Insert(10);
    hash.Insert(15);
    hash.Insert(20);
    hash.Insert(25);
    hash.Insert(30);
end.
